src/libwalletqt/PassphraseHelper.h

```diff
@@ -0,0 +1,149 @@
+
+#ifndef MONERO_GUI_PASSPHRASEHELPER_H
+#define MONERO_GUI_PASSPHRASEHELPER_H
+
+#include <QtGlobal>
+#include <wallet/api/wallet2_api.h>
+#include <QMutex>
+#include <QPointer>
+#include <QWaitCondition>
+#include <QMutex>
+
+/**
+ * Implements component responsible for showing entry prompt to the user,
+ * typically Wallet / Wallet manager.
+ */
+class PassphraseHandler {
+public:
+    virtual void onWalletPassphraseNeeded(bool onDevice) = 0;
+};
+
+/**
+ * Implements receiver of the passphrase responsible for passing it back to the wallet,
+ * typically wallet listener.
+ */
+class PassphraseDelegate {
+public:
+    virtual void onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort) = 0;
+};
+
+class PassphraseHelper {
+public:
+    PassphraseHelper(PassphraseHandler * handler=nullptr): m_handler(handler) {};
+    PassphraseHelper(const PassphraseHelper & h): PassphraseHelper(h.m_handler) {};
+    Monero::optional<std::string> onDevicePassphraseRequest(bool & on_device);
+    void onWalletPassphraseNeeded(bool on_device);
+    void onPassphraseEntered(const QString &passphrase, bool enter_on_device, bool entry_abort);
+
+private:
+    PassphraseHandler * m_handler;
+    QWaitCondition m_cond_pass;
+    QMutex m_mutex_pass;
+    QString m_passphrase;
+    bool m_passphrase_abort;
+    bool m_passphrase_on_device;
+
+};
+
+// Compatibility with lower Qt versions
+#if (QT_VERSION >= QT_VERSION_CHECK(5, 12, 0))
+#include <QScopeGuard>
+#else
+#ifndef QSCOPEGUARD_H
+#define QSCOPEGUARD_H
+#include <QtCore/qglobal.h>
+
+QT_BEGIN_NAMESPACE
+template <typename T, typename U = T>
+Q_DECL_RELAXED_CONSTEXPR T qExchange(T &t, U &&newValue)
+{
+    T old = std::move(t);
+    t = std::forward<U>(newValue);
+    return old;
+}
+
+template <typename F> class QScopeGuard;
+template <typename F> QScopeGuard<F> qScopeGuard(F f);
+template <typename F>
+class
+#if QT_HAS_CPP_ATTRIBUTE(nodiscard)
+// Q_REQUIRED_RESULT can be defined as __warn_unused_result__ or as [[nodiscard]]
+// but the 1st one has some limitations for example can be placed only on functions.
+Q_REQUIRED_RESULT
+#endif
+QScopeGuard
+{
+public:
+    QScopeGuard(QScopeGuard &&other) noexcept
+        : m_func(std::move(other.m_func))
+        , m_invoke(qExchange(other.m_invoke, false))
+    {
+    }
+
+    ~QScopeGuard()
+    {
+        if (m_invoke)
+            m_func();
+    }
+
+    void dismiss() noexcept
+    {
+        m_invoke = false;
+    }
+
+private:
+    explicit QScopeGuard(F &&f) noexcept
+        : m_func(std::move(f))
+    {
+    }
+
+    Q_DISABLE_COPY(QScopeGuard)
+
+    F m_func;
+    bool m_invoke = true;
+    friend QScopeGuard qScopeGuard<F>(F);
+};
+
+
+template <typename F>
+#if QT_HAS_CPP_ATTRIBUTE(nodiscard)
+Q_REQUIRED_RESULT
+#endif
+QScopeGuard<F> qScopeGuard(F f)
+{
+    return QScopeGuard<F>(std::move(f));
+}
+
+QT_END_NAMESPACE
+
+#endif // QSCOPEGUARD_H
+#endif
```

Comment:
Please move `QScopeGuard` stuff to a separate header in `src/qt`

Reply:
Done, pls check and resolve when ok.

Reply:
Seems fine to me now. Feel free to `Resolve` it.